### PR TITLE
[12.0][MIG] website_sale_charge_payment_fee: Migration to 12.0

### PR DESCRIPTION
**Pull request by TimLai125**
I re-created this PR from https://github.com/OCA/e-commerce/pull/374 to preserve commit history as requested.

### CODE REVIEW

File: website_sale_charge_payment_fee/static/src/js/website_sale_fee.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+            // See website_sale/static/src/js/website_sale_payment.js :
+            // .find("input[name='acquirer']:checked").click();
+            // Without this check, redirect would be performed right after page load
+            if (
+                selected_acquirer_id != null &&
+                payment_id != selected_acquirer_id
+            ) {
+                $pay_button.prop("disabled", true);
+                window.location.href = "/shop/payment?payment_fee_id=" + payment_id;
+            }
```
Comment (yostashiro) [inline website_sale_charge_payment_fee/static/src/js/website_sale_fee.js]: Do we need this for V12? `website_sale/static/src/js/website_sale_payment.js` looks quite different.

File: website_sale_charge_payment_fee/controllers/main.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+from odoo import http
+from odoo.http import request
+
+from odoo.addons.website_sale.controllers.main import WebsiteSale
+
+
+class WebsiteSaleFee(WebsiteSale):
+    @http.route()
+    def payment(self, **post):
```
Comment (yostashiro) [inline website_sale_charge_payment_fee/controllers/main.py:14]: This method doesn't seem to be called when I try selecting an acquirer in my local environment.

File: website_sale_charge_payment_fee/static/src/js/website_sale_fee.js [hunk 3]
```diff
@@ -1,29 +1,23 @@
...
-        var selected_acquirer_id = $selected_acquirer_id.val();
-        $payment_method.find("input[name='pm_id']").click(function(ev) {
-            var payment_id = $(ev.currentTarget)
-                .val()
+    require("web.dom_ready");
+    require("web.ajax");
+    require("web.core");
+
+    $(".o_payment_form").each(function() {
+        var payment_form = this;
```
Comment (yostashiro) [inline website_sale_charge_payment_fee/static/src/js/website_sale_fee.js]:
```suggestion
```

File: website_sale_charge_payment_fee/static/src/js/website_sale_fee.js [hunk 4]
```diff
@@ -1,29 +1,23 @@
...
+        var selected_acquirer = $("input[name='selected_acquirer_id']").val();
+
+        $(payment_form).on("click", 'input[name="pm_id"]', function() {
```
Comment (yostashiro) [inline website_sale_charge_payment_fee/static/src/js/website_sale_fee.js]:
```suggestion
        $(this).on("click", 'input[name="pm_id"]', function() {
```

File: website_sale_charge_payment_fee/static/src/js/website_sale_fee.js [hunk 5]
```diff
@@ -1,29 +1,23 @@
...
+            var clicked_acquirer = $(this)
+                .prop("value")
                 .substring(5);
```
Comment (yostashiro) [inline website_sale_charge_payment_fee/static/src/js/website_sale_fee.js]:
```suggestion
                .substring(5);  // Represents the form id.
```

File: website_sale_charge_payment_fee/static/src/js/website_sale_fee.js [hunk 6]
```diff
@@ -1,29 +1,23 @@
...
-            // .find("input[name='acquirer']:checked").click();
-            // Without this check, redirect would be performed right after page load
-            if (
-                selected_acquirer_id != null &&
-                payment_id != selected_acquirer_id
-            ) {
-                $pay_button.prop("disabled", true);
-                window.location.href = "/shop/payment?payment_fee_id=" + payment_id;
+            if (selected_acquirer !== clicked_acquirer) {
+                $("#o_payment_form_pay").addClass("disabled");
```
Comment (yostashiro) [inline website_sale_charge_payment_fee/static/src/js/website_sale_fee.js]: Is this check needed at all? I didn't experience any strange behavior in my local environment disabling this condition.

If this is still needed, we should also revive the comments to make the intention clear.

File: website_sale_charge_payment_fee/models/payment_acquirer.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    _inherit = "payment.acquirer"
+
+    charge_fee = fields.Boolean(
+        "Fee charged to customer",
+        help="An extra fee line will be added to online order when using this "
+        "payment method",
+    )
+    charge_fee_description = fields.Text("Fee Description")
+    charge_fee_product_id = fields.Many2one("product.product", string="Fee Product")
+    charge_fee_tax_ids = fields.Many2many("account.tax", string="Taxes")
```
Comment (yostashiro) [inline website_sale_charge_payment_fee/models/payment_acquirer.py]:
```suggestion
```
I think we should delete the taxes here, although I am not sure about the intention by the original author. IMO the taxes should be proposed from the product, just like any other products.

File: website_sale_charge_payment_fee/models/sale_order.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+            self.env["sale.order.line"].create(
+                {
+                    "order_id": self.id,
+                    "payment_fee_line": True,
+                    "product_id": acquirer.charge_fee_product_id.id,
+                    "product_uom": acquirer.charge_fee_product_id.uom_id.id,
+                    "name": acquirer.charge_fee_description,
+                    "price_unit": price,
+                    "product_uom_qty": 1,
+                    "tax_id": [(6, 0, [t.id for t in acquirer.charge_fee_tax_ids])],
```
Comment (yostashiro) [inline website_sale_charge_payment_fee/models/sale_order.py]:
```suggestion
```
This line is a problem when you try to adjust the taxes based on the fiscal position.

File: website_sale_charge_payment_fee/views/payment_acquirer_views.xml [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+                                <field
+                                    name="charge_fee_percentage"
+                                    class="oe_inline"
+                                />%
+                            </div>
+                            <field
+                                name="charge_fee_tax_ids"
+                                widget="many2many_tags"
+                                attrs="{'invisible': [('charge_fee', '=', False)]}"
+                            />
```
Comment (yostashiro) [inline website_sale_charge_payment_fee/views/payment_acquirer_views.xml]:
```suggestion
```

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: website_sale_charge_payment_fee/models/payment_acquirer.py (hunk 7 above)
Comment (yostashiro) [reply website_sale_charge_payment_fee/models/payment_acquirer.py]: @eLBati Can you perhaps comment on this please?

Comment (eLBati) [reply website_sale_charge_payment_fee/models/payment_acquirer.py]: @yostashiro I don't remember if this wanted to cover any specific requirement, but I don't think so.
I agree with your approach.
Thanks

Comment (yostashiro) [reply website_sale_charge_payment_fee/models/payment_acquirer.py]: @eLBati Thanks for your comment!

**Comment by TimLai125** [comment]
@pedrobaeza It seems that the travis is failling but it is nothing related to the module that I am working on. Can you advise what I should do to proceed the PR?

**Comment by pedrobaeza** [comment]
Are you sure you are working with a rebase over latest 12.0 branch?

Doing changes on the QWeb templates can affect other module tests as the selectors of the tours are based on the DOM. You can try to include an isolated build to check if that's the problem. See .travis.yml to see how it's done.

File: website_sale_charge_payment_fee/models/sale_order.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+                    order.order_line.filtered("payment_fee_line").mapped(
+                        "price_subtotal"
+                    )
+                )
+            else:
+                order.amount_payment_fee = sum(
+                    order.order_line.filtered("payment_fee_line").mapped("price_total")
+                )
+
+    def update_fee_line(self, acquirer):
```
Comment (yostashiro) [inline website_sale_charge_payment_fee/models/sale_order.py:50]:
```suggestion
    def update_fee_line(self, acquirer):
        self.ensure_one()
```

File: website_sale_charge_payment_fee/views/website_sale_templates.xml [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                            t-field="acq.charge_fee_percentage"
+                        />
+                        %
+                    </t>
+                    )
+                </span>
+            </t>
+        </xpath>
+    </template>
+    <template id="total" name="Payment Fee" inherit_id="website_sale.total">
```
Comment (yostashiro) [inline website_sale_charge_payment_fee/views/website_sale_templates.xml]:
```suggestion
    <template id="payment_fee" name="Payment Fee" inherit_id="website_sale.total">
```

File: website_sale_charge_payment_fee/models/sale_order.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    amount_payment_fee = fields.Monetary(
+        compute="_compute_amount_payment_fee",
+        digits=0,
+        string="Payment Fee Amount",
+        store=True,
+        track_visibility="always",
+    )
```
Comment (yostashiro) [inline website_sale_charge_payment_fee/models/sale_order.py:20]: Would be nice to mention that the field (and logic) are made consistent with amount_delivery from website_sale_delivery.

**Review by yostashiro — CHANGES_REQUESTED** [review]
Almost there. The design is much cleaner now.

**Comment by yostashiro** [comment]
Runbot has some warnings but I can't see the logs here: https://runbot.odoo-community.org/runbot/build/3453221

Could someone advise what we should do? @Yajo @pedrobaeza @rafaelbn @simahawk @gfcapalbo

**Review by yostashiro — APPROVED** [review]
LGTM, except for the warnings in Runbot which I'm not clear what to do about.

**Review by lehoangan — APPROVED** [review]

**Pull request merged**